File: libmuscle/python/libmuscle/mpp_client.py
from typing import List, Optional

import msgpack
from ymmsl import Reference

from libmuscle.mcp.protocol import RequestType
from libmuscle.mcp.transport_client import TransportClient
from libmuscle.mcp.type_registry import transport_client_types


class MPPClient:
    """A client that connects to an MPP server.

    This client connects to a peer to retrieve messages. It uses an MCP
    Transport to connect.
    """
    def __init__(self, locations: List[str]) -> None:
        """Create an MPPClient for the given peer.

        The client will connect to the peer on one of its locations. It
        tries the most efficient protocol first. Once connected, it can
        request messages from any component and port represented by it.

        Args:
            locations: The peer's location strings
        """
        client = None       # type: Optional[TransportClient]
        for ClientType in transport_client_types:
            for location in locations:
                if ClientType.can_connect_to(location):
                    try:
                        client = ClientType(location)
                        break
                    except Exception:
                        pass
            if client:
                break
        else:
            raise RuntimeError('Failed to connect')

        self._transport_client = client

    def receive(self, receiver: Reference) -> bytes:
        """Receive a message from a port this client connects to.

        Args:
            receiver: The receiving (local) port.

        Returns:
            The received message.
        """
        request = [RequestType.GET_NEXT_MESSAGE.value, str(receiver)]
        encoded_request = msgpack.packb(request, use_bin_type=True)
        return self._transport_client.call(encoded_request)

    def close(self) -> None:
        """Closes this client.

        This closes any connections this client has and/or performs
        other shutdown activities.
        """
        self._transport_client.close()

File: libmuscle/cpp/src/libmuscle/mpp_client.cpp
#include "libmuscle/mpp_client.hpp"

#include "libmuscle/data.hpp"
#include "libmuscle/mcp/protocol.hpp"
#include "libmuscle/mcp/data_pack.hpp"
#include "libmuscle/mcp/tcp_transport_client.hpp"

#include <memory>
#include <stdexcept>
#include <vector>


using libmuscle::_MUSCLE_IMPL_NS::Data;
using libmuscle::_MUSCLE_IMPL_NS::DataConstRef;
using libmuscle::_MUSCLE_IMPL_NS::mcp::ProfileData;
using libmuscle::_MUSCLE_IMPL_NS::mcp::TcpTransportClient;
using ymmsl::Reference;


namespace libmuscle { namespace _MUSCLE_IMPL_NS {

MPPClient::MPPClient(std::vector<std::string> const & locations) {
    try_connect_<TcpTransportClient>(locations);
    if (!transport_client_)
        throw std::runtime_error("Could not connect to peer");
}

std::tuple<std::vector<char>, ProfileData> MPPClient::receive(
        Reference const & receiver, mcp::TimeoutHandler *timeout_handler)
{
    auto request = Data::list(
            static_cast<int>(RequestType::get_next_message),
            std::string(receiver));

    msgpack::sbuffer sbuf;
    // TODO: can we put in an 8-byte dummy value here, which we
    // can then overwrite after encoding with the length?
    msgpack::pack(sbuf, request);

    return transport_client_->call(sbuf.data(), sbuf.size(), timeout_handler);
}

void MPPClient::close() {
    transport_client_->close();
}


template <class ClientType> void MPPClient::try_connect_(
        std::vector<std::string> const & locations
) {
    for (auto const & location : locations) {
        if (ClientType::can_connect_to(location)) {
            try {
                transport_client_ = std::make_unique<ClientType>(location);
                break;
            }
            catch (std::runtime_error const & e) {}
        }
    }
}

} }

